src/io/flutter/inspector/InspectorService.java

```diff
@@ -255,6 +257,80 @@ private void onVmServiceReceived(String streamId, Event event) {
     return invokeServiceMethodDaemonNoGroup(methodName, params);
   }
 
+  /*
+   * We make a best guess for the pub directory based on the the root directory
+   * of the first non artifical widget in the tree.
+  */
+  public CompletableFuture<String> inferPubRootDirectoryIfNeeded(){
+    ObjectGroup group = createObjectGroup("temp");
+    DiagnosticsNode root = group.getRoot(FlutterTreeType.widget).join();
+
+    if (root == null) {
+      // No need to do anything as there isn't a valid tree (yet?).
+      group.dispose();
+      return null;
+    }
+
+    ArrayList<DiagnosticsNode> children = root.getChildren().join();
+
+    if (children != null && children.isEmpty() == false) {
+      // There are already widgets identified as being from the summary tree so
+      // no need to guess the pub root directory.
+      return null;
+    }
+
+    ArrayList<DiagnosticsNode> allChildren =
+      group.getChildren(root.getDartDiagnosticRef(), false, null).join();
```

Comment:
avoid using join. Likely you are freezing the IntelliJ UI thread which is where this method was probably called from. Use the async methods on the CompletableFuture class instead to make this method Async like the other methods on this class. That helps keep the logic consistent on the Dart and Java sides although it makes the Java code a bit painful to write.

Reply:
changed to thenComposeAsync.